### PR TITLE
C++: Improvements to the cpp/toctou-race-condition query

### PR DESCRIPTION
**Pull request by geoffw0**
Improvements to the `cpp/toctou-race-condition` query.  The main change is disallowing results from file operations such as `fopen`, `rename` etc from counting as the "check", so now the check must be output from a `stat`, `access` or similar call, whilst the "use" can still be a wide range of file operations.  Most of the results this removes are either clear FPs or of dubious interest, so this is a big improvement to precision.

I've made a couple of other improvements that widen the results a little (not as much as the above narrows them), in hopefully uncontroversial ways.
 - fix the logic for use of `stat` arguments so that it actually works in most cases.
 - recognize `_wfsopen`.
 - recognize the second argument of `rename` (some implementations will overwrite an existing file).

LGTM diff query results here: https://lgtm.com/query/4593284742579958544/

I haven't upgraded the precision of this query to `high` as yet, but I'm thinking about it.  The main remaining class of FP / weak result IMO is where `stat` is being used to check a file exists before doing something that would perhaps fail if it didn't; or `stat` is checking another not-so-security-relevant property.  These are still race conditions, but not necessarily security issues, so I think I'd *rather* not report them but I don't have a strong opinion.  On the flip side, this is pretty much what the examples on https://cwe.mitre.org/data/definitions/367.html do and what the SAMATE tests do, so perhaps I'm being too strict.

### CODE REVIEW

**Comment by geoffw0** [comment]
Here are a couple of real world toctou CVEs:
 - https://bugs.gentoo.org/214403
 - https://alioth-lists-archive.debian.net/pipermail/secure-testing-team/2008-June/001672.html

The attacks often involve the program checking some security relevant property of the file (such as the user having permission to access it), then the attacker switches that file for a symlink to the real target, then the program operates on that target using its privilege but without having checked the property on it.

**Comment by geoffw0** [comment]
Hmm, there are some regressions in the old internal tests.  I'll make a PR to accept the changes, but I'm beginning to think there's a second case we really want to permit as well: any check, followed by a highly sensitive operation such as `chmod`.  The motivating case is an `fopen(filename, "w")` followed by granting a permission on that (believed to be newly written) file using `chmod`.  I can imagine other combinations such as `rename` followed by `chown` on the renamed file that would be dangerous if the file were switched with a symlink or something.

**Comment by geoffw0** [comment]
Pushed changes and internal PR.  New diff query: https://lgtm.com/query/6665823716934953527/ .  There aren't many changes, but we get back what I believe are good results in a few places such as stellar/stellar-core.

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql [hunk 1]
```diff
@@ -26,28 +26,29 @@ import semmle.code.cpp.controlflow.Guards
...
-      name = "open" or
-      name = "freopen" or
-      name = "_open" or
-      name = "_wopen" or
-      name = "_wfopen"
-    ) and
+    name =
+      [
+        "remove", "unlink", "rmdir", "rename", "fopen", "open", "freopen", "_open", "_wopen",
+        "_wfopen", "_wfsopen"
```
Comment (rdmarsh2) [inline cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql:32]: There's also a non-wide `_fsopen` variant

**Review by rdmarsh2 — COMMENTED** [review]
Results and code LGTM other than the one nitpick.

File: cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql]: Thanks, the more the merrier (done).

**Review by rdmarsh2 — APPROVED** [review]
LGTM, and I'm merging now to avoid submodule merge conflicts. Do we want to bump the precision to `high` as a separate PR?

**Comment by geoffw0** [comment]
> Do we want to bump the precision to high as a separate PR?

I'll review the results again first, but yes.

**Pull request merged**